README.md
Elgg News Plugin
================

![Elgg 2.3](https://img.shields.io/badge/Elgg-2.3-orange.svg?style=flat-square)

Elgg plugin for posting news and announcements by administrators, group owners or news staff.

## Features

This plugin offers admins the options:

1. Post news items
2. Add other Elgg entities as news items
3. Option to determine more users who can post news/announcements
4. Option to change the default news icon and featured news icon in settings
5. Customizable news list view

As an example for 2nd case, if there is blog post by any user, administrator can add this as news post. So it will be displayed in list of news but if users click on this item, will redirected to original blog post. The same action could apply to any post like pages, bookmarks, videos etc. 

Also administrator can allow to group owners for posting new/announcements inside the group only.


## How to use a customizable news list view
If need to have a different view for listing news, e.g. in a custom front page, you could use the following code:

```php
echo elgg_view('amapnews/custom_list_view', array(
    'entities' => $entities,    // list of news entities, previously retrieved
    'read_more' => true,        // set true if want to add a "Read more" link for each news item
    'item_class' => '',         // set a custom class on news items, so it could be customized through CSS
    'photo_size' => 'custom',   // set the size of the news photo, see more details below
    'photo_cover' => false,     // set true if want to use the news as cover image, otherwise it will be displayed inline with title and intro
    'photo_class' => '',        // set a custom class on news photo
));
```

You can see an example about how to use this view at:
- amapnews/views/default/resources/amapnews/custom_list_view.php

This sample view will be accessible at http://www.YOURCOMMUNITYURL.com/news/custom_list

Especially about photo size, you can use any of the predefined photo size, which are:
```php
elgg_set_config('amapnews_photo_sizes', array(
    'tiny' => array('w' => 25, 'h' => 25, 'square' => TRUE, 'upscale' => FALSE),
    'small' => array('w' => 40, 'h' => 40, 'square' => TRUE, 'upscale' => FALSE),
    'medium' => array('w' => 100, 'h' => 100, 'square' => TRUE, 'upscale' => FALSE),
    'large' => array('w' => 150, 'h' => 150, 'square' => TRUE, 'upscale' => FALSE),
    'master' => array('w' => 215, 'h' => 215, 'square' => TRUE, 'upscale' => FALSE),
    'default' => array('w' => 1200, 'h' => 1200, 'square' => FALSE, 'upscale' => FALSE),
));
```

Or you can set a custom photo size (width and height) in plugin settings, so for each photo upload on news item, it will be saved in this size too.

## How to change the default icon
Site administrators are able to select in plugin settings the icons to use for:
1. Default news icon
2. Featured news icon

If need to add a custom icon, just uploaded it to folders **mod/amapnews/graphics/icons** for default news icon or to **mod/amapnews/graphics/featured** for featured news icon. All file uploads will be available for selection in plugin settings.


CHANGES.md
### 2.3.9	(March 2017)
- Added option to restrict setting news as featured by administrators only

### 2.3.8	(March 2017)
- Added option to set custom size on news photos
- Added a customizable news list view
- Added option in settings to change the default news icon and featured news icon
- Fixed: when saving a news post, the user is redirected to list of news

### 2.3.8	(March 2017)
- Added compatibility with Elgg 2.3.x
- Added line break on summary input/output 
- Improved view news posts on widgets
- Fixed issue on tags field
- Fixed issue on user icon

### 2.0.6
- fixed: bug while trying to post/delete news in groups as non-admin user
- fixed: small issues

### 2.0.5
- added: option to show news photos on walled garden sites
- fixed: set featured/unfeatured is allowed only to entity owner or admins
- fixed: posts with comments set to Off cannot be commented on river
- fixed: composer file

### 2.0.4	(June 2016)
- Added compatibility with Elgg 2.0
- Added option to set a news item as featured
- Added a featured news widget 
- Added option to upload photos on news item
- News posts are now likeable 
- Plugin now can be installed via composer
- Several code improvements

### 1.11.3	(May 2015)
- Added option to determine users who can post news/announcements (thanks to Pasley70 for contribution)
- Added group widget
- Small bugs fixes

### 1.8.2	(Dec 2014)
- Small bugs fixes

### 1.8.1	(Nov 2014)
- Added option for group owners to post announcements inside group only
- Small bugs fixes

### 1.8.0	(15 Nov 2014)
Initial release 








start.php
<?php
/**
 * Elgg News plugin
 * @package amapnews
 */
 
require_once(dirname(__FILE__) . '/lib/hooks.php');
require_once(dirname(__FILE__) . '/lib/widgets.php');

elgg_register_event_handler('init', 'system', 'amapnews_init');

define('AMAPNEWS_GENERAL_YES', 'yes');	// general purpose string for yes
define('AMAPNEWS_GENERAL_NO', 'no');	// general purpose string for no

/**
 * amapnews plugin initialization functions.
 */
function amapnews_init() {
 	
    // register a library of helper functions
    elgg_register_library('elgg:amapnews', elgg_get_plugins_path() . 'amapnews/lib/amapnews.php');
    
    // register extra css
    elgg_extend_view('elgg.css', 'amapnews/amapnews.css');
        
    // Register entity_type for search
    elgg_register_entity_type('object', Amapnews::SUBTYPE);
    
    // Site navigation
    $item = new ElggMenuItem('amapnews', elgg_echo('amapnews:menu'), 'news');
    elgg_register_menu_item('site', $item); 
    
    // add option to all site entities for adding to news
    elgg_register_plugin_hook_handler('register', 'menu:entity', 'amapnews_entity_menu_setup', 400);

    // add option for admin to add/remove user from news-staff
    elgg_register_plugin_hook_handler("register", "menu:user_hover", "news_staff_user_hover_menu_hook");
    
    // Register menu item to an ownerblock. It is used to  register news menu item to groups
    elgg_register_plugin_hook_handler('register', 'menu:owner_block', 'amapnews_owner_block_menu');
    
    // allow to be liked
    elgg_register_plugin_hook_handler('likes:is_likable', 'object:amapnews', 'Elgg\Values::getTrue');
    
    // Register a URL handler for news
    elgg_register_plugin_hook_handler('entity:url', 'object', 'amapnews_set_url');
    
    // We don't want people commenting on news posts in the river
    elgg_register_plugin_hook_handler('permissions_check:comment', 'object', 'amapnews_comment_override');    
    
    // Register a page handler, so we can have nice URLs
    elgg_register_page_handler('news', 'amapnews_page_handler');
    elgg_register_page_handler('amapnews', 'amapnews_page_handler');

    // extend group main page 
    elgg_extend_view('groups/tool_latest', 'amapnews/group_module');
    
    // add the group news tool option
    add_group_tool_option('amapnews', elgg_echo('amapnews:group:enable'), true);   
	
    // loads the widgets
    amapnews_widgets_init();
    
    // set photo sizes for news posts
    elgg_set_config('amapnews_photo_sizes', array(
        'tiny' => array('w' => 25, 'h' => 25, 'square' => TRUE, 'upscale' => FALSE),
        'small' => array('w' => 40, 'h' => 40, 'square' => TRUE, 'upscale' => FALSE),
        'medium' => array('w' => 100, 'h' => 100, 'square' => TRUE, 'upscale' => FALSE),
        'large' => array('w' => 150, 'h' => 150, 'square' => TRUE, 'upscale' => FALSE),
        'master' => array('w' => 215, 'h' => 215, 'square' => TRUE, 'upscale' => FALSE),
        'default' => array('w' => 1200, 'h' => 1200, 'square' => FALSE, 'upscale' => FALSE),
    ));    
    
    // Register actions
    $action_path = elgg_get_plugins_path() . 'amapnews/actions/amapnews';
    elgg_register_action('amapnews/add', "$action_path/add.php");    
    elgg_register_action('amapnews/delete', "$action_path/delete.php");
    elgg_register_action('amapnews/staff', "$action_path/staff.php", "admin");
    elgg_register_action('amapnews/set_featured', "$action_path/set_featured.php", "admin");    
}

/**
 *  Dispatches amapnews pages.
 *
 * @param array $page
 * @return bool
 */
function amapnews_page_handler($page) {
    elgg_push_breadcrumb(elgg_echo('amapnews'), 'news');
    
    if (!isset($page[0])) {
        $page[0] = 'all';
    }    
    $vars = array();
    $vars['page'] = $page[0];	

    $resource_vars = array();
    switch ($page[0]) {
        case "add":
            $resource_vars['guid'] = elgg_extract(1, $page);
            echo elgg_view_resource('amapnews/add', $resource_vars);
            break;
        
        case "photo_view":
            $resource_vars['guid'] = elgg_extract(1, $page);
            $resource_vars['size'] = elgg_extract(2, $page);
            $resource_vars['tu'] = elgg_extract(3, $page);
            $resource_vars['ia'] = elgg_extract(4, $page);
            echo elgg_view_resource('amapnews/photo_view', $resource_vars);
            break;        
            
        case "add_existed":
            $resource_vars['cguid'] = elgg_extract(1, $page);
            echo elgg_view_resource('amapnews/add_existed', $resource_vars);
            break;            
            
        case "edit":
            $resource_vars['guid'] = elgg_extract(1, $page);
            echo elgg_view_resource('amapnews/edit', $resource_vars);
            break;
            
        case "view":
            $resource_vars['guid'] = elgg_extract(1, $page);
            echo elgg_view_resource('amapnews/view', $resource_vars);
            break;   
            
        case "owner":
            echo elgg_view_resource('amapnews/owner');
            break;   
            
        case "group":
            group_gatekeeper();
            echo elgg_view_resource('amapnews/owner');
            break;
            
        case "all":
            echo elgg_view_resource('amapnews/all');
            break;
        
        case "custom_list":
            // this is a demonstration on how to get a custom view, e.g. in front page
            $resource_vars['limit'] = elgg_extract(1, $page);
            echo elgg_view_resource('amapnews/custom_list_view', $resource_vars);
            break;        
            
        default:
            echo elgg_view_resource('amapnews/all');
            return false;
    }

    elgg_pop_context();
    return true;
}

?>
